### PR TITLE
feat: Implement REVERT opcode

### LINKED ISSUE #82: Opcode: REVERT

### PR DESCRIPTION
**Pull request by Danikim01**
closes #82 
This PR implements the revert opcode

### CODE REVIEW

File: src/codegen/operations.rs [hunk 1]
```diff
@@ -1848,11 +1850,108 @@ fn codegen_return<'c>(
 
     extend_memory(op_ctx, &ok_block, required_size)?;
 
-    op_ctx.write_result_syscall(&ok_block, offset, size, location);
+    let reamining_gas = get_remaining_gas(context, &ok_block)?;
```
Comment (MegaRedHand) [inline src/codegen/operations.rs:1853]: typo: reamining -> remaining

File: src/codegen/operations.rs [hunk 2]
```diff
@@ -1848,11 +1850,108 @@ fn codegen_return<'c>(
...
+        .into();
+
+    let required_size = ok_block
+        .append_operation(arith::addi(offset, size, location))
+        .result(0)?
+        .into();
+
+    extend_memory(op_ctx, &ok_block, required_size)?;
+
+    let reamining_gas = get_remaining_gas(context, &ok_block)?;
```
Comment (MegaRedHand) [inline src/codegen/operations.rs:1925]: Same typo here.

File: src/syscall.rs [hunk 3]
```diff
@@ -22,33 +22,106 @@ use melior::ExecutionEngine;
...
+            _ => Self::Default,
+        }
+    }
+}
+
+#[derive(Debug)]
+pub enum ExecutionResult {
+    Success {
+        /// The offset and size in [`Self::memory`] corresponding to the EVM return data.
+        /// It's [`None`] in case there's no return data
```
Comment (MegaRedHand) [inline src/syscall.rs:50]: This comment is stale

File: src/syscall.rs [hunk 4]
```diff
@@ -22,33 +22,106 @@ use melior::ExecutionEngine;
...
+        return_data: Vec<u8>,
+        gas_remaining: u64,
+    },
+    Revert {
+        /// The offset and size in [`Self::memory`] corresponding to the EVM return data.
+        /// It's [`None`] in case there's no return data
```
Comment (MegaRedHand) [inline src/syscall.rs:56]: This one too

File: src/syscall.rs [hunk 5]
```diff
@@ -22,33 +22,106 @@ use melior::ExecutionEngine;
...
+        execution_result: u8,
+    ) {
+        self.return_data = Some((offset as usize, bytes_len as usize));
+        self.gas_remaining = Some(remaining_gas);
+        match ExitStatusCode::from_u8(execution_result) {
+            ExitStatusCode::Return => self.exit_status = Some(ExitStatusCode::Return),
+            ExitStatusCode::Revert => self.exit_status = Some(ExitStatusCode::Revert),
+            ExitStatusCode::Error => self.exit_status = Some(ExitStatusCode::Error),
+            ExitStatusCode::Default => self.exit_status = Some(ExitStatusCode::Default),
+        }
```
Comment (MegaRedHand) [inline src/syscall.rs:124]: We can reduce this to: `self.exit_status = Some(ExitStatusCode::from_u8(execution_result));`

File: src/syscall.rs [hunk 6]
```diff
@@ -72,17 +145,19 @@ impl SyscallContext {
 
 pub mod symbols {
     pub const WRITE_RESULT: &str = "emv_mlir__write_result";
+    pub const REVERT_RESULT: &str = "emv_mlir__store_revert_result";
```
Comment (MegaRedHand) [inline src/syscall.rs:148]: We can remove this now.

Comment (Danikim01) [reply src/syscall.rs:148]: fixed

File: src/syscall.rs (hunk 5 above)
Comment (Danikim01) [reply src/syscall.rs:124]: fixed

File: tests/operations.rs [hunk 7]
```diff
@@ -24,13 +27,46 @@ fn run_program_assert_result_with_gas(
         .compile(&program, &output_file)
         .expect("failed to compile program");
 
-    let executor = Executor::new(&module);
+    let engine = ExecutionEngine::new(module.module(), 0, &[], false);
```
Comment (MegaRedHand) [inline tests/operations.rs:30]: These changes seem to revert the move to using `Executor`

Comment (Danikim01) [reply tests/operations.rs:30]: what do you mean by this?

File: Cargo.toml [hunk 8]
```diff
@@ -8,6 +8,7 @@ license = "LICENSE"
 keywords = ["ethereum", "evm"]
 
 [dependencies]
+assert_matches = "1.5.0"
```
Comment (MegaRedHand) [inline Cargo.toml:11]: Please move to dev-dependencies section

**Review by MegaRedHand — APPROVED** [review]
Great work!

**Pull request merged**